### ISSUE TITLE
Porting to Fabric 1.17 and up?

### ISSUE THREAD

**Issue by James103** [yeelp/Scaling-Feast#119]
**Is your feature request related to a problem? Please describe.**
Currently, there appears to be no way to use the mod as it is on Minecraft versions 1.17 and up.
1.12 is no longer supported by most mod and mod pack creators, and a declining fraction of players use it.

**Describe the feature you'd like**
It would be great to port this to the latest Minecraft versions, including 1.17, 1.18 (when it comes out), and later.
Ideally, this would be ported to [Fabric](https://fabricmc.net/) for easier maintenance.
_A Forge may be acceptable as well depending on the circumstances._

**Describe why this feature belongs in Scaling Feast**
By porting this mod to modern versions such as 1.17 and up, you open up the mod to being used by more players, especially those that play on modern versions of Minecraft.
By porting this mod to Fabric, you make it easier to maintain and more accessible to people.

**Describe alternatives you've considered**
You could try to find a mod that does the same thing as what this mod does, however, there is likely no such mod.
You could try to recreate the mod server-side-only using data packs or functions, but you would not be able to recreate all of the functionality.

**Comment by yeelp** [comment]
I'm not sure about the feasibility, Scaling Feast relies heavily on AppleCore for a large majority of its features, and leaves handling max hunger to AppleCore, and AppleCore is stuck in 1.12 (check out the discussion [here](https://github.com/squeek502/AppleCore/issues/142)). A Forge PR has been made that adds AppleCore's features to Forge, but it hasn't been merged last I checked. Porting to newer versions is effectively stalled until that change is merged. It may be possible to port to newer versions without AppleCore, but that would be a rewrite instead of a port, which would take much more time (and not all features may be able to make it).

As for Fabric, I don't know how it differs from Forge. What makes Fabric a better Forge alternative for modern versions?

**Comment by James103** [comment]
See the following posts for example:
https://forums.minecraftforge.net/topic/87328-forge-fabric-difference/
https://www.reddit.com/r/feedthebeast/comments/e57sps/what_is_the_functional_difference_between_forge/
https://www.reddit.com/r/feedthebeast/comments/j7buq9/whats_the_difference_between_forge_and_fabric_and/
... and many of the other links found on https://duckduckgo.com/?q=difference+between+fabric+and+forge

To summarize, Fabric is lightweight and allows mods to more easily mod the vanilla code directly. It loads faster and is smaller / more efficient compared to Forge. In addition, Fabric can update to newer Minecraft versions much faster compared to Forge, mainly as a result of it being more lightweight and reduced development overhead.

For example, a mod created for the Fabric mod loader can mixin into the HungerManager (and many other classes in the Minecraft source code) directly, but that may not be the case for Forge.

**Comment by yeelp** [comment]
If that's the case, then ScalingFeast may only be possible for modern versions as a Fabric exclusive, so I can see that as a possibility, but potentially a challenging one. Something to tackle after #113

**Comment by James103** [comment]
> ScalingFeast may only be possible for modern versions as a Fabric exclusive

At the very least, until MinecraftForge/MinecraftForge#7266 is merged and updated to modern versions.

**Comment by James103** [comment]
Regarding porting Scaling Feast to Fabric, what I would like to have happen is that (on Fabric) the Scaling Feast mod itself contains the new mechanics, effects, blocks and items etc, while a separate library mod tentatively called "Hunger Attributes" (`hunger-attributes`) has the actual hunger attributes themselves.

This should allow other mods and datapacks, not just Scaling Feast, to make use of the new attributes to modify various numbers about a player's hunger system.

AppleSkin (or a fork thereof) may need to be updated to support the new library mod, allowing the hunger and saturation bars to be at their proper size with respect to the corresponding attributes.

<details>

<summary> List of attributes to be implemented by Hunger Attributes </summary>

### Max Hunger (`hunger_attributes:max_hunger`)
* Modifies the maximum hunger level a player can have.
* Default value: 20, same as in vanilla.
* Example 1: Setting this to 40 allows the player's food level to be 2x greater compared to vanilla. Client-side, their hunger bar will gain an additional row, have colored overlays or just scale to the higher maximum (with each half-shank representing 2 hunger).
* Example 2: Setting this to 10 limits the player's food level to half that of vanilla. Client side, their hunger bar will either show only 5 shank slots or scale to the lower maximum (with each full shank representing 1 hunger).
* Minimum value: 1 (a player needs some hunger or else they starve)

### Max Saturation Ratio (`hunger_attributes:max_saturation_ratio`)
* Multiplies the maximum saturation level a player can have (compared to their hunger).
* Default value: 1.0, same as in vanilla.
* Example 1: Setting this to 2.0 allows the player to have twice as much saturation compared to their hunger.
* Example 2: Setting this to 0.5 caps the player's saturation at half their hunger level.
* Minimum value: 0, which does not allow the player to gain starvation.

### Exhaustion Cap (`hunger_attributes:exhaustion_cap`)
* Modifies the maximum exhaustion level a player can ever have.
* Default value: 40.0, same as in vanilla.
* Example 1: Setting this to 1000.0 allows the player to reach 1000.0 exhaustion at once, which if left unchecked will drain 250 food points.
* Example 2: Setting this to 8.0 allows the player to lose a maximum of 2 food points from a single action, even if said action would cost more.
* Minimum value: 0, which does not allow the player to gain exhaustion.

### Max Exhaustion (`hunger_attributes:max_exhaustion`)
* Modifies the exhaustion level required for a player to lose a food point (first saturation then hunger).
* Default value: 4.0, same as in vanilla.
* Example 1: Setting this to 10.0 requires the player to reach 10 exhaustion before losing a food point.
* Example 2: Setting this to 1.0 causes the player to lose 1 food point every 1.0 exhaustion (effectively four times faster than vanilla).
* Minimum value: 0.0001 (or another small positive value) - note: a value of 0 here would divide by zero, causing a crash or an infinite drain of 1 food point per tick.

### Food Efficiency (`hunger_attributes:food_efficiency`)
* Modifies the amount of exhaustion gained from all sources, effectively making food last longer (or shorter).
* Default value: 1.0, same as in vanilla.
* Example 1: Applying a "+40% Food Efficiency" modifier causes the player to accumulate exhaustion $1 + 0.4 = 1.4\times$ slower.
* Example 2: Applying a "-60% Food Efficiency" modifier causes the player to accumulate exhaustion $\frac{1}{1 - 0.6} = \frac{1}{0.4} = 2.5\times$ faster.
* Minimum value: 0.0001 (or another small positive value) - note: a value of 0 here would divide by zero, causing a crash or forcing exhaustion to be at its hardcap every player action.

</details>

**Comment by yeelp** [comment]
Library mod is a great idea. Currently, Fabric seems to be the only way forward for Scaling Feast.

We return to the age old question of how to handle AppleSkin. A fork *could* be made, but I may stick to doing it in a similar manner to how it was handled in Scaling Feast.

I'm not sure how much I like the arbitrary minimum values for exhaustion; there's gotta be some reasonable and logical minimum value or a different way of representing the attribute that has a logical minimum value. Food efficiency should already be designed in a way that avoids divide by zero errors (since it's a percentage increase, and thus has no minimum value) so an option for max exhaustion should exist too

**Comment by James103** [comment]
If the behavior for food efficiency in the current version of Scaling Feast described in https://github.com/yeelp/Scaling-Feast/issues/81 still how you want the attribute to work in its port?
> When food efficiency is greater than 1, divide exhaustion gained by the food efficiency (this is normal behavior). When food efficiency is less than 1, however, take 2 minus the food efficiency value and multiply exhaustion gained by the result.

If so, the attribute for 'max exhaustion' may need to get the same treatment. I've updated the descriptions of the corresponding proposed attributes to match.

**Comment by yeelp** [comment]
Yeah, it makes sense to leave the food efficiency attribute functionally as is. I think it's implemented well enough.
